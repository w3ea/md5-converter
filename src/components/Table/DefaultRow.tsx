import { FC } from 'react';
import cls from 'clsx';
import style from '../../assets/styles/components/table/column.module.scss';

const DefaultRow: FC = () => (
    <tr>
        <td className={cls(style.column, style.column__text)} data-label='Text'>
            <span>
                Welcome to MyMd5Converter. Use the form above to get started
            </span>
        </td>
        <td className={cls(style.column, style.column__md5)} data-label='MD5'>
            <span>
                f5c4a3b160d818bdb8e0190b9041ba44
            </span>
        </td>
    </tr>
);

export default DefaultRow;