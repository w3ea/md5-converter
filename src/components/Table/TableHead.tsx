import { FC } from 'react';
import { useDispatch } from 'react-redux';
import { removeAllItem } from '../../store/tableListSlice';
import style from '../../assets/styles/components/table/column.module.scss';

type Props = {
    existUserList: boolean
};

const TableHead: FC<Props> = ({ existUserList }) => {
    const dispatch = useDispatch();

    const onRemoveAll = () => {
        // Check if we are in the client not server
        if (typeof window !== 'undefined') {
            window.confirm('Are you sure to delete all items?') && dispatch(removeAllItem()); // eslint-disable-line no-alert
        }
    };

    return (
        <thead className='lg:bg-gray-800 lg:border-b lg:border-blue-100'>
            <tr className='hidden lg:flex w-full'>
                {existUserList && (
                    <th className={style.column__number}>
                        #
                    </th>
                )}
                <th className={style.column__text}>
                    Text
                </th>
                <th className={style.column__md5}>
                    MD5
                </th>
                {existUserList && (
                    <th className={style.column__action}>
                        <button
                            className='tw-table__head__remove-all-button'
                            type='button'
                            onClick={onRemoveAll}
                            title='Remove all'
                        >
                            <svg className='w-6' xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 24 24' stroke='currentColor'>
                                <path strokeLinecap='round' strokeLinejoin='round' strokeWidth={2} d='M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16' />
                            </svg>
                        </button>
                    </th>
                )}
            </tr>
        </thead>
    );
};

export default TableHead;