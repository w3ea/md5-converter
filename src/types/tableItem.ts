export type TableItem = {
    id: string,
    originalText: string,
    MD5Hash: string
};