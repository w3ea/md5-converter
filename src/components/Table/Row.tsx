import { FC } from 'react';
import { useDispatch } from 'react-redux';
import { motion } from 'framer-motion';
import cls from 'clsx';
import { removeItem } from '../../store/tableListSlice';
import { TableItem } from '../../types/tableItem';
import style from '../../assets/styles/components/table/column.module.scss';

const rowAnimate = {
    hidden: { opacity: 0, x: -20 },
    visible: { opacity: 1, x: 0, transition: { duration: 0.3 } },
    exit: { opacity: 0, x: 20, transition: { duration: 0.3 } }
};

type Props = {
    rowNumber: number,
    item: TableItem
};

const Row: FC<Props> = ({ rowNumber, item }) => {
    const dispatch = useDispatch();

    const onRemove = () => {
        // Check if we are in the client not server
        if (typeof window !== 'undefined') {
            window.confirm('Are you sure to delete this item?') && dispatch(removeItem(item.id)); // eslint-disable-line no-alert
        }
    };

    return (
        <motion.tr initial='hidden' animate='visible' exit='exit' variants={rowAnimate}>
            <td className={cls(style.column, style.column__number)} data-label='#'>
                <span>
                    {rowNumber}
                </span>
            </td>
            <td className={cls(style.column, style.column__text)} data-label='Text' data-testid='text'>
                <span>
                    {item.originalText}
                </span>
            </td>
            <td className={cls(style.column, style.column__md5)} data-label='MD5'>
                <span>
                    {item.MD5Hash}
                </span>
            </td>
            <td className={cls(style.column, style.column__action)} data-label='Action'>
                <div className='py-2.25 px-2 lg:py-3.5 lg:px-4 mx-auto'>
                    <button
                        data-testid='remove-item-button'
                        className='tw-table__row__remove-item'
                        type='button'
                        onClick={onRemove}
                        title='Remove item'
                    >
                        <svg className='w-5 md:w-5.5' xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 24 24' stroke='currentColor'>
                            <path strokeLinecap='round' strokeLinejoin='round' strokeWidth={2} d='M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16' />
                        </svg>
                        <p className='text-3.5 font-bold lg:hidden'>
                            Delete
                        </p>
                    </button>
                </div>
            </td>
        </motion.tr>
    );
};

export default Row;