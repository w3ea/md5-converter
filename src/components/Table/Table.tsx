import { FC } from 'react';
import { useSelector } from 'react-redux';
import { AnimatePresence } from 'framer-motion';
import { tableListSelector } from '../../store/tableListSlice';
import TableHead from './TableHead';
import DefaultRow from './DefaultRow';
import Row from './Row';
import style from '../../assets/styles/components/table/table.module.scss';

const Table: FC = () => {
    const tableList = useSelector(tableListSelector);
    const existUserList = tableList.length > 0;

    return (
        <div className='tw-table'>
            <table className={style.table} summary='The table list of converted text contains the row number, text, md5, and action (delete item) columns'>
                <TableHead existUserList={existUserList} />
                <tbody>

                    {existUserList && (
                        <AnimatePresence initial={false}>
                            {tableList?.map((item, index) => (
                                <Row key={item.id} item={item} rowNumber={index + 1} />
                            ))}
                        </AnimatePresence>
                    )}

                    {! existUserList && <DefaultRow />}
                </tbody>
            </table>
        </div>
    );
};

export default Table;