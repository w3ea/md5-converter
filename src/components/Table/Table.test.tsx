import { render, fireEvent, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';

import { configureStore, DeepPartial } from '@reduxjs/toolkit';
import Table from './Table';
import TableListSlice, { addNewItem } from '../../store/tableListSlice';
import { TableItem } from '../../types/tableItem';

window.confirm = jest.fn(() => true);

const preloaded: DeepPartial<{ tableList: TableItem[]; }> = {
    tableList: [
        {
            id: '31b28e4b-1bc2-40c4-a13c-ad3b6cad3297',
            originalText: 'foo',
            MD5Hash: 'acbd18db4cc2f85cedef654fccc4a4d8'
        },
        {
            id: 'c9319be5-e2b6-4b6e-b3b6-806e1af6e02a',
            originalText: 'bar',
            MD5Hash: '37b51d194a7513e45b56f6524f2d51f2'
        }
    ]
};

const store = configureStore({ reducer: { tableList: TableListSlice }, preloadedState: preloaded });

const renderTable = (): RenderResult => render(
    <Provider store={store}>
        <Table />
    </Provider>
);

describe('<Table />', () => {
    it('should match table snapshot', () => {
        const { asFragment } = renderTable();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should add new item to table', () => {
        store.dispatch(addNewItem({ originalText: 'test', MD5Hash: 'acbd18db4cc2f85cedef654fccc4a4d8' }));

        const { getByText } = renderTable();

        expect(getByText('test')).toBeInTheDocument();
    });

    it('should add new item to the first row', () => {
        store.dispatch(addNewItem({ originalText: 'test', MD5Hash: 'acbd18db4cc2f85cedef654fccc4a4d8' }));

        const { getAllByTestId } = renderTable();

        const firstRowText = getAllByTestId('text')[0].textContent;

        expect(firstRowText).toBe('test');
    });

    it('should remove item from table', () => {
        const { getAllByTestId } = renderTable();

        const removeButton = getAllByTestId('remove-item-button')[0];
        fireEvent.click(removeButton);

        expect(window.confirm).toBeCalled();
    });
});