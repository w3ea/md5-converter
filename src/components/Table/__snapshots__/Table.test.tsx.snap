// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should match table snapshot 1`] = `
<DocumentFragment>
  <div
    class="tw-table"
  >
    <table
      class="table"
      summary="The table list of converted text contains the row number, text, md5, and action (delete item) columns"
    >
      <thead
        class="lg:bg-gray-800 lg:border-b lg:border-blue-100"
      >
        <tr
          class="hidden lg:flex w-full"
        >
          <th
            class="column__number"
          >
            #
          </th>
          <th
            class="column__text"
          >
            Text
          </th>
          <th
            class="column__md5"
          >
            MD5
          </th>
          <th
            class="column__action"
          >
            <button
              class="tw-table__head__remove-all-button"
              title="Remove all"
              type="button"
            >
              <svg
                class="w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="sr-only"
              >
                Remove all items
              </span>
            </button>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          style="opacity: 1; transform: none;"
        >
          <td
            class="column column__number"
            data-label="#"
          >
            <span>
              1
            </span>
          </td>
          <td
            class="column column__text"
            data-label="Text"
            data-testid="text"
          >
            <span>
              foo
            </span>
          </td>
          <td
            class="column column__md5"
            data-label="MD5"
          >
            <span>
              acbd18db4cc2f85cedef654fccc4a4d8
            </span>
          </td>
          <td
            class="column column__action"
            data-label="Action"
          >
            <div
              class="py-2.25 px-2 lg:py-3.5 lg:px-4 mx-auto"
            >
              <button
                class="tw-table__row__remove-item"
                data-testid="remove-item-button"
                title="Remove item"
                type="button"
              >
                <svg
                  class="w-5 md:w-5.5"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
                <p
                  class="text-3.5 font-bold lg:hidden"
                >
                  Delete
                </p>
              </button>
            </div>
          </td>
        </tr>
        <tr
          style="opacity: 1; transform: none;"
        >
          <td
            class="column column__number"
            data-label="#"
          >
            <span>
              2
            </span>
          </td>
          <td
            class="column column__text"
            data-label="Text"
            data-testid="text"
          >
            <span>
              bar
            </span>
          </td>
          <td
            class="column column__md5"
            data-label="MD5"
          >
            <span>
              37b51d194a7513e45b56f6524f2d51f2
            </span>
          </td>
          <td
            class="column column__action"
            data-label="Action"
          >
            <div
              class="py-2.25 px-2 lg:py-3.5 lg:px-4 mx-auto"
            >
              <button
                class="tw-table__row__remove-item"
                data-testid="remove-item-button"
                title="Remove item"
                type="button"
              >
                <svg
                  class="w-5 md:w-5.5"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
                <p
                  class="text-3.5 font-bold lg:hidden"
                >
                  Delete
                </p>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
